### PR TITLE
417 add support to send invoices by webservices to portuguese at

### LINKED ISSUE #417: Add support to send invoices by webservices to Portuguese AT
Portuguese law Decreto-Lei n.º 198/2012  requires businesses to send invoicing data by one of two ways: 
 - real-time transmission by way of SOAP webservices
 - Manual submission of a SAF-T file.   

https://dre.pt/dre/legislacao-consolidada/decreto-lei/2012-106458692
```
Artigo 3.º
Comunicação dos elementos das faturas
1 - As pessoas, singulares ou coletivas, que se encontrem sujeitas às regras de emissão de faturação em território português nos termos do artigo 35.º-A do Código do IVA e aqui pratiquem operações sujeitas a IVA, são obrigadas a comunicar à Autoridade Tributária a Aduaneira (AT), por transmissão eletrónica de dados, os elementos das faturas emitidas nos termos do Código do IVA, bem como os elementos dos documentos que possibilitem a conferência de mercadorias ou de prestação de serviços e recibos, por uma das seguintes vias:
a) Por transmissão eletrónica de dados em tempo real;
b) Por transmissão eletrónica de dados, mediante remessa de ficheiro normalizado estruturado com base no ficheiro SAF-T (PT), criado pela Portaria n.º 321-A/2007, de 26 de março, na sua redação atual;
c) Por inserção direta no Portal das Finanças;
d) (Revogada.)
```

Exporting SAF-T is already supported in billy. But for some applications it would be better to use real time transmission using SOAP webservices. 

Documentation is provided in: https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/Pages/default.aspx
 - [SAF-T PT (Standard Audit File for Tax purposes) - Versão Portuguesa ](https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/SAFT_PT/Paginas/news-saf-t-pt.aspx )
 - https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/Faturacao/Fatcorews/Documents/Fatcorews.wsdl
 - Operations provided:
     - RegisterInvoice - Registo de um novo documento comercial
     - ChangeInvoiceStatus - Alteração do estado do documento comercial
     - DeleteInvoice - Eliminação de documentos comerciais
     - RegisterWork - Registo de um novo documento conferência
     - ChangeWorkStatus - Alteração do estado do documento de conferência
     - DeleteWork - Eliminação de documentos de conferência
     - RegisterPayment - Registo de um novo documento de recibo
     - ChangePaymentStatus - Alteração do estado do recibo
     - DeletePayment - Eliminação de documentos de recibos
 - https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/Faturacao/Fatcorews/Documents/Comunicacao_dos_elementos_dos_documentos_de_faturacao.pdf
 - SSL Client certificate for testing: https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/Faturacao/Documents/TesteWebservices.zip 
 - Testing certificates for SOAP header authentication encryption https://faturas.portaldasfinancas.gov.pt/factemipf_static/java/certificados.zip
 - https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/Faturacao/Documents/ChaveCifraPublicaAT2025.zip

### PR DESCRIPTION
**Pull request by froque**
fixes #417 

When to invoke this service and dealing with retries is left to the client application and not managed by billy. 

No custom exceptions are being thrown. All checked exceptions are being wrapped in `RuntimeException`.

### CODE REVIEW

File: billy-portugal/src/main/resources/TesteWebservices.pfx
Comment (froque) [inline billy-portugal/src/main/resources/TesteWebservices.pfx:1]: Expires: 04/09/2023

File: billy-portugal/src/main/resources/saPubKey.jks
Comment (froque) [inline billy-portugal/src/main/resources/saPubKey.jks:1]:
```
❯ keytool -list -v -storepass saKeyPubPass -keystore billy-portugal/src/main/resources/saPubKey.jks 
Keystore type: JKS
Keystore provider: SUN

Your keystore contains 2 entries

Alias name: sapubkey.prod
Creation date: 06/07/2020
Entry type: trustedCertEntry

Owner: CN=Chave Cifra Publica AT, OU=Sistemas de Informacao, O=Autoridade Tributaria e Aduaneira, L=Lisboa, ST=Lisboa, C=PT
Issuer: CN=DGITA Issuing CA2, DC=ritta, DC=local
Serial number: 2f00001c56506e04bf31e7e8bf000100001c56
Valid from: Tue Jun 23 18:50:12 WEST 2020 until: Fri Jun 23 18:50:12 WEST 2023
Certificate fingerprints:
	 SHA1: AE:88:CA:F0:68:56:6F:15:EC:6A:85:84:17:D9:9D:56:7B:DE:8A:77
	 SHA256: DE:FD:74:C1:64:F1:ED:71:76:E7:EE:1E:F9:14:58:5D:7E:71:51:69:10:30:05:41:5E:B9:4C:7E:EB:29:F4:C3
Signature algorithm name: SHA256withRSA
Subject Public Key Algorithm: 2048-bit RSA key
Version: 3

Extensions: 

#1: ObjectId: 1.2.840.113549.1.9.15 Criticality=false
0000: 30 35 30 0E 06 08 2A 86   48 86 F7 0D 03 02 02 02  050...*.H.......
0010: 00 80 30 0E 06 08 2A 86   48 86 F7 0D 03 04 02 02  ..0...*.H.......
0020: 00 80 30 07 06 05 2B 0E   03 02 07 30 0A 06 08 2A  ..0...+....0...*
0030: 86 48 86 F7 0D 03 07                               .H.....


#2: ObjectId: 1.3.6.1.4.1.311.21.10 Criticality=false
0000: 30 0E 30 0C 06 0A 2B 06   01 04 01 82 37 0A 03 04  0.0...+.....7...


#3: ObjectId: 1.3.6.1.4.1.311.21.7 Criticality=false
0000: 30 2E 06 26 2B 06 01 04   01 82 37 15 08 82 8D 9D  0..&+.....7.....
0010: 39 82 8E EC 2A 84 FD 93   13 82 F6 D5 6E 82 99 BF  9...*.......n...
0020: 73 52 83 99 91 25 82 A5   AD 7A 02 01 64 02 01 1A  sR...%...z..d...


#4: ObjectId: 1.3.6.1.5.5.7.1.1 Criticality=false
AuthorityInfoAccess [
  [
   accessMethod: caIssuers
   accessLocation: URIName: ldap:///CN=DGITA%20Issuing%20CA2,CN=AIA,CN=Public%20Key%20Services,CN=Services,CN=Configuration,DC=ritta,DC=local?cACertificate?base?objectClass=certificationAuthority
]
]

#5: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
KeyIdentifier [
0000: C4 4B 73 8D D5 60 96 7B   BE A2 7B 7D 25 F0 B7 E2  .Ks..`......%...
0010: 35 6F FC 1E                                        5o..
]
]

#6: ObjectId: 2.5.29.31 Criticality=false
CRLDistributionPoints [
  [DistributionPoint:
     [URIName: ldap:///CN=DGITA%20Issuing%20CA2,CN=SW500210,CN=CDP,CN=Public%20Key%20Services,CN=Services,CN=Configuration,DC=ritta,DC=local?certificateRevocationList?base?objectClass=cRLDistributionPoint]
]]

#7: ObjectId: 2.5.29.37 Criticality=false
ExtendedKeyUsages [
  1.3.6.1.4.1.311.10.3.4
]

#8: ObjectId: 2.5.29.15 Criticality=false
KeyUsage [
  Key_Encipherment
  Data_Encipherment
]

#9: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: ED B3 BA 8C AE E7 F0 C0   FE 61 D0 73 28 3A 1F 38  .........a.s(:.8
0010: F3 4D FC 3F                                        .M.?
]
]



*******************************************
*******************************************


Alias name: sapubkey.testes
Creation date: 14/02/2020
Entry type: trustedCertEntry

Owner: CN=Teste SA Acesso, OU=NAGC, O=Autoridade Tributaria e Aduaneira, L=Lisboa, ST=Lisboa, C=PT
Issuer: CN=DGITA Issuing CA1, DC=ritta, DC=local
Serial number: 120000212d7c5a904a244efabb00030000212d
Valid from: Wed Sep 25 15:25:51 WEST 2019 until: Thu Sep 24 15:25:51 WEST 2020
Certificate fingerprints:
	 SHA1: 3A:D2:04:24:0B:B7:18:07:D7:01:E1:1A:2D:BF:BA:6D:B9:D8:3D:9B
	 SHA256: 81:02:EB:A5:0A:10:19:C5:6D:E1:47:6F:90:43:89:49:B4:A3:65:87:77:8F:C5:3C:1F:78:06:F5:67:26:04:F2
Signature algorithm name: SHA256withRSA
Subject Public Key Algorithm: 2048-bit RSA key
Version: 3

Extensions: 

#1: ObjectId: 1.3.6.1.4.1.311.21.10 Criticality=false
0000: 30 18 30 0A 06 08 2B 06   01 05 05 07 03 01 30 0A  0.0...+.......0.
0010: 06 08 2B 06 01 05 05 07   03 02                    ..+.......


#2: ObjectId: 1.3.6.1.4.1.311.21.7 Criticality=false
0000: 30 2E 06 26 2B 06 01 04   01 82 37 15 08 82 8D 9D  0..&+.....7.....
0010: 39 82 8E EC 2A 84 FD 93   13 82 F6 D5 6E 82 99 BF  9...*.......n...
0020: 73 52 86 E9 D1 3B 81 A2   BA 0D 02 01 64 02 01 72  sR...;......d..r


#3: ObjectId: 1.3.6.1.5.5.7.1.1 Criticality=false
AuthorityInfoAccess [
  [
   accessMethod: caIssuers
   accessLocation: URIName: ldap:///CN=DGITA%20Issuing%20CA1,CN=AIA,CN=Public%20Key%20Services,CN=Services,CN=Configuration,DC=ritta,DC=local?cACertificate?base?objectClass=certificationAuthority
]
]

#4: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
KeyIdentifier [
0000: 31 93 9D CE 6A E2 D1 EE   C1 A7 9D F4 AD A9 74 8C  1...j.........t.
0010: 40 FC FC C2                                        @...
]
]

#5: ObjectId: 2.5.29.31 Criticality=false
CRLDistributionPoints [
  [DistributionPoint:
     [URIName: ldap:///CN=DGITA%20Issuing%20CA1(1),CN=SW500212,CN=CDP,CN=Public%20Key%20Services,CN=Services,CN=Configuration,DC=ritta,DC=local?certificateRevocationList?base?objectClass=cRLDistributionPoint]
]]

#6: ObjectId: 2.5.29.37 Criticality=false
ExtendedKeyUsages [
  serverAuth
  clientAuth
]

#7: ObjectId: 2.5.29.15 Criticality=false
KeyUsage [
  DigitalSignature
  Key_Encipherment
]

#8: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 91 C5 41 53 8B 83 AC D8   85 96 93 DC BB F8 1D 1D  ..AS............
0010: E9 92 08 92                                        ....
]
]



*******************************************
*******************************************

```

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/AuthenticationHandler.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+        KeyStore ks = KeyStore.getInstance("JKS");
+
+        try (InputStream fis = webserviceKeys.getKeystore().toURL().openStream()) {
+            ks.load(fis, webserviceKeys.getKeystorePassword().toCharArray());
+            return ks.getCertificate(webserviceKeys.getKeyAlias()).getPublicKey();
+        }
+    }
+
+    private static final SimpleDateFormat TIMESTAMP_FORMATER = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S'Z'");
```
Comment (thenmaster) [inline billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/AuthenticationHandler.java:175]: Constants should be at the start of the file

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+    public Client(Config config, URL url, WebserviceCredentials webserviceCredentials, WebserviceKeys webserviceKeys,
+            SslClientCertificate sslClientCertificate) {
+        this.config = config;
+        this.url = url;
+        this.webserviceCredentials = webserviceCredentials;
+        this.webserviceKeys = webserviceKeys;
+        this.sslClientCertificate = sslClientCertificate;
+    }
+
+    Response sendInvoice(PTApplication application, PTGenericInvoice ptGenericInvoice){
```
Comment (thenmaster) [inline billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java:94]: Any particular reason this method(and the delete methods as well) are package private? Won't this limit their usability?

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+        documentTotals.setGrossTotal(ptGenericInvoice.getAmountWithTax());
+        invoiceData.setDocumentTotals(documentTotals);
+        registerInvoiceRequest.setInvoiceData(invoiceData);
+        return registerInvoiceRequest;
+    }
+
+    static class LineSummaryValue {
+        final BigDecimal lineAmount;
+        final BigDecimal taxPercentage;
+        final BigDecimal totalTaxAmount;
```
Comment (thenmaster) [inline billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java:245]: Spacing

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+        if (ptGenericInvoice.isCancelled()) {
+            invoiceStatus.setInvoiceStatus("A");
+        } else if (ptGenericInvoice.isBilled()) {
+            invoiceStatus.setInvoiceStatus("F");
+        } else if (ptGenericInvoice.isSelfBilled()) {
+            invoiceStatus.setInvoiceStatus("S");
+        } else {
+            invoiceStatus.setInvoiceStatus("N");
+        }
+        invoiceStatus.setInvoiceStatusDate(formatDateTime(ptGenericInvoice.getDate()));
```
Comment (thenmaster) [inline billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java:448]: This date doesn't account for the changes in upstream that added timezones to businesses and local dates to SAFT and export logics.

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/SslClientCertificate.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with billy portugal (PT Pack). If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.premiumminds.billy.portugal.services.export.webservice;
+
+import java.net.URI;
+
+public class SslClientCertificate {
+    private final URI path;
```
Comment (thenmaster) [inline billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/SslClientCertificate.java:24]: Spacing

File: billy-portugal/src/test/java/com/premiumminds/billy/portugal/services/export/webservice/ClientTest.java [hunk 6]
```diff
@@ -0,0 +1,703 @@
...
+
+        publish = Endpoint.publish(bindingURI, webService);
+    }
+
+    @AfterEach
+    public void tearDown(){
+        publish.stop();
+    }
+
+    static class LineSummaryKey {
```
Comment (thenmaster) [inline billy-portugal/src/test/java/com/premiumminds/billy/portugal/services/export/webservice/ClientTest.java:146]: You defined an identical class in the Client implementation. You could define it in a separate file and then reuse the code in both the client and this test

**Review by thenmaster — CHANGES_REQUESTED** [review]

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/AuthenticationHandler.java (hunk 1 above)
Comment (froque) [reply billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/AuthenticationHandler.java]: changed

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java (hunk 3 above)
Comment (froque) [reply billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java:249]: changed

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java (hunk 2 above)
Comment (froque) [reply billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java]: nice catch. thanks
changed

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/SslClientCertificate.java (hunk 5 above)
Comment (froque) [reply billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/SslClientCertificate.java:25]: changed

File: billy-portugal/src/test/java/com/premiumminds/billy/portugal/services/export/webservice/ClientTest.java (hunk 6 above)
Comment (froque) [reply billy-portugal/src/test/java/com/premiumminds/billy/portugal/services/export/webservice/ClientTest.java]: moved to proper class

File: billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java (hunk 4 above)
Comment (froque) [reply billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/Client.java]: updated

**Review by thenmaster — APPROVED** [review]

**Pull request merged**